api/services/coinsService.js

```diff
@@ -0,0 +1,232 @@
+const { getMultipleKeyDetails } = require('../../db/redis');
+const { getMetadataForPids, getClient } = require('../../db/elastic');
+const { initCacheWatcher, getLocalCache, updateLocalCache } = require('../cache/metadataCache');
+
+initCacheWatcher();
+
+async function getCoinMetadata(query) {
+  let { pid } = query;
+  let pidArray = [];
+
+  if (Array.isArray(pid)) {
+    pidArray = pid.map((p) => p.trim().toLowerCase()).filter(Boolean);
+  } else if (typeof pid === 'string') {
+    pidArray = pid.split(',').map((p) => p.trim().toLowerCase()).filter(Boolean);
+  }
+
+  if (!pidArray.length) {
+    throw new Error("Parameter 'pid' is required for metadata.");
+  }
+
+  const localMetadataCache = getLocalCache();
+  const resultMap = {};
+  const missingPids = [];
+
+  for (const p of pidArray) {
+    if (localMetadataCache[p]) {
+      resultMap[p] = localMetadataCache[p];
+    } else {
+      missingPids.push(p);
+    }
+  }
+
+  if (missingPids.length > 0) {
+    const fetchedMap = await getMetadataForPids(missingPids);
+    for (const [missingPid, data] of Object.entries(fetchedMap)) {
+      localMetadataCache[missingPid] = data;
+      resultMap[missingPid] = data;
+    }
+    updateLocalCache(localMetadataCache);
+  }
+
+  for (const p of pidArray) {
+    if (!resultMap[p]) {
+      resultMap[p] = null;
+    }
+  }
+
+  return resultMap;
+}
+
+/**
+ * Retrieve the current coin data from Redis. 
+ * If `withMetadata=true`, attach metadata using the same logic as getCoinMetadata.
+ */
+async function getCoinsService(pidString, options = {}) {
+  const { withTTL = false, withMetadata = false } = options;
+
+  const pids = pidString
+    .split(',')
+    .map(p => p.trim().toLowerCase())
```

Comment:
instead of .toLowerCase, we should using normalizePid here, it takes care of edge cases. 

Also, we need to maintain a map between input pid & normalized value, as the response structure matches input